packages/core/contracts/insured-bridge/RateModelStore.sol

```diff
@@ -0,0 +1,29 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "../common/implementation/MultiCaller.sol";
+
+/**
+ * @title Maps rate model objects to L1 token.
+ * @dev This contract is designed to be queried by off-chain relayers that need to compute realized LP fee %'s before
+ * submitting relay transactions to a BridgePool contract. Therefore, this contract does not perform any validation on
+ * the shape of the rate model, which is stored as a string to enable arbitrary data encoding via a stringified JSON
+ * object. This leaves this contract unopionated on the parameters within the rate model, enabling governance to adjust
+ * the structure in the future.
+ */
+contract RateModelStore is Ownable, MultiCaller {
```

Comment:
Hm this is interesting, so store a `startBlock` and have a public `getRateModel` that returns the rate model depending on the current block?

Reply:
I think this is a good idea because it allows the owner of this contract (i.e. the admin of Across) to give off-chain clients time to prepare for an update.

Reply:
Yeah, something like that. I'm not sure how we might implement this, but the feature seems useful. And it's easy to not use by just setting the block to some value pretty soon in the future.

Reply:
@mrice32 I implemented it PTAL sorry I forgot to mention

Reply:
nice! there is still a situation where this can become problematic that we'd need to think about. if the rate model switches _after_ the bot has pulled it down but _before_ the bot sends transactions it could use the wrong rate model on it's sends. I don't think this is a huge risk but could pose problems. The right solution to avoid this I think is to pause deposits for a few mins on either side of updating the rate model.

if we use pausing then I don't think this extra logic that you added is needed. wdyt?

Reply:
I don't quite understand the attack, you're talking about the very rare edge case where the rate model changes in between the bot querying the contract state `getRateModel` and submitting the transaction? So this is a ~5 minute window that the rate model could change?

I do think pausing would work though

Reply:
well it's less of an attack more just to point out that I think this logic of `startBlock` is redundant (and somewhat lacking) in preventing the issue where the rate model changes under the user (during their execution loop, between when they query the rate model and when they submit transactions it changes).

my point really is: we should not include `startBlock` in the contract and rather pause deposits for a few mins before and after changing rate models. I think that is safer and more predictable.

wdyt?

Reply:
Hmm, does that mean for each deposit box on N networks we need to send N pause transactions and then N unpause transactions? I wish we could manage this in the RateModelStore

Reply:
So @mrice32 @chrismaree it sounds like we have three choices for implementing rate model upgrades. These choices have different ways to combat the edge case where clients are using an out-dated rate model compared to the relayer bots that will dispute them:

1) Allow owner to set a `startBlock` for the `rateModel` to become active. Setting a `startBlock` in the future combined with pausing deposits on the L2 deposit boxes prevents clients' rate models from being out of sync with relayer bots.
2) Same as 1, but don't pause deposits on L2 deposit boxes. Live with the low probability that clients' rate models are out of sync with relayer bots.
2) Pause L2 deposit boxes until new `rateModel` is stored in `RateModelStore` contract. This route would enable us to remove the `startBlock` logic from the `RateModelStore`.

Reply:
@chrismaree I think I agree with your critique here, but I may arrive at a different conclusion:
- I agree that the start block doesn't operationally fix the issue of the rate model changing since the bot will only see that new rate model _after_ it changes. The only advantage it offers is that the bot might be able to pull down the future rate model and make sure it can still process it as a way of warning the runner early if the schema has changed.
- I don't think pausing deposits is a sustainable solution (or really solves the problem). It's bad from a UX perspective and it makes the upgrades really painful because we have to wait for the cross-chain message to take effect before sending the upgrade, then waiting for another cross chain message to unpause. This could result in 20 minutes+ of downtime. Even if this were done, what happens if a deposit pre-pause were relayed in the middle of the upgrade (same block, for instance). The bot would still have to figure out what the correct rate model is.

The right solution is that we must have a way to determine the rate model _at any time_. The identifier must be resolvable for _any_ relay.
- The rate model updates instantaneously when the transaction is mined. No startBlock required -- as you've pointed out @chrismaree, the startBlock doesn't solve any problems.
- For each relay, you take the quote timestamp and look at the closest mainnet block that has a timestamp <= that quote timestamp.
- You query your rate model for that relay _at that block_.
- This means that if your quote timestamp matches the timestamp for a block where the rate model changes, you take the new rate model. If it is less than that, you take the rate model before that change.

I think this would simplify the contract and the UMIP, and the logic would be pretty similar to our existing fee computation logic (we do the same thing for querying utilization). Thoughts?

Reply:
@mrice32 that's a great solution! Simple and powerful. It has some consequences though. This means you can't validate relays older than 30 mins without an archival node. This is no different than the current assumptions though where you need to get utilisation at the deposit time to get realized LP fees.

Reply:
Nice thing though that this look back requirement does not really impact the front end

Reply:
>  This means you can't validate relays older than 30 mins without an archival node. This is no different than the current assumptions though where you need to get utilisation at the deposit time to get realized LP fees.

Yeah this doesn't change our assumptions from the bot's perspective, although it makes this contract less useful in a standalone context. But it would never be used standalone

Reply:
I also like @mrice32 's solution because it keeps the contract simple and moves complexity to the off-chain implementation of `getRateModel`

Reply:
I've reverted this PR back to the original implementation where the `updateRateModel` function simply sets a new rate model. There is no "start block" logic.